### DISCUSSION TITLE
Implementing the TomTom Traffic APIs

### DISCUSSION THREAD

**Discussion by ZeljkoBG** [osmandapp/OsmAnd#20103] [Ideas]
Is that possible?
https://www.tomtom.com/products/traffic-apis/

or Yandex Traffic API

**Comment by MarioSwitch** [comment]
I hope this will be implemented, it would be a HUGE feature addition!
And as [this doc presenting new features of 4.8 version](https://osmand.net/docs/versions/future-android) says, this is one of the "possible" task, there is a chance this will be implemented in a few weeks/months!

**Reply by ZeljkoBG** [reply]
It has already been implemented, but only in two countries

**Reply by MarioSwitch** [reply]
I've heard that, you're right, but in which countries and how is this displayed?

**Reply by MarioSwitch** [reply]
Just noticed this in the 4.9 doc source file 👀

![image](https://github.com/user-attachments/assets/89114f16-cb84-4b6b-811d-6ee0f946c0e6)

So it's for Germany and the Netherlands...

**Comment by TobiPeterG** [comment]
It would be nicer if there was an open traffic API, not a proprietary one.

**Reply by MarioSwitch** [reply]
Yes, but I don't know if this exists... 🤔

**Reply by TobiPeterG** [reply]
No, not yet ;)

**Reply by TobiPeterG** [reply]
A comment I made on another issue:

I would love to see an open traffic API. And I guess other apps like organic maps would also be very interested in this.
How I could imagine this to work, is something similar to this:
https://github.com/opentraffic/otv2-platform

Have a Server that aggregates the traffic data and this could then also allow online routing, e.g. using Valhalla, as opentraffic already uses it.
So maybe take that as a base and expand it for the features OSMAnd needs?

**Comment by mariush444** [comment]
[routing.xml has a section called obstacle](https://github.com/osmandapp/OsmAnd-resources/blob/bfa257b412ecc843ec38a08cdf43826d0e8180ae/routing/routing.xml#L1013) dedicated to tracking the cost associated with points.

So I guess points can be used to mark accidents, traffic jams to improve navigation.

Would be possible to dynamicaly change xml and recalculate route if needed.

the question is how to share that points or points like "temporary avoid the road"

**Reply by MarioSwitch** [reply]
TomTom APIs provides JSON elements, with penalties in minutes. I think there is no need to change XML, just add these penalties to the total time, and if possible, recalculate route when changing.

As APIs updates traffic every 30 seconds, it could be possible, for example, to recalculate route on every update **IN THE BACKGROUND**, and if the newly calculated route is different from displayed route, change displayed route with a toast "Route has been changed due to traffic conditions" or something like that...

It could be also great if route summary shows traffic lates (e.g. "5h 13m + 24m traffic = 5h 37m"), a widget indicating total traffic lates until destination, and support Start/End of traffic jams, accidents, roadworks in the navigation widget (to indicate "Traffic jam in 8km, +14min" with new icons)

**Reply by MarioSwitch** [reply]
![image](https://github.com/user-attachments/assets/3c5a7f70-67c9-476d-8227-df06d1e067aa)

Example of screen, when there are 14 min traffic lates in 47.3 km, and roadworks 167 km ahead 😊

**Comment by yekibud** [comment]
Does anyone know if the Traffic implementation for Germany/Netherlands uses the TomTom API, or something else?  And what are the obstacles to implementing it for USA and other countries?  I'm not an Android dev, but I'm a full-stack/cloud dev and would like to contribute, especially to support availability for the US and make the feature (plugin?) available for optional open/crowd-sourced traffic APIs (such as regional or federated servers based on the OpenTraffic platform that @TobiPeterG suggested).  Any suggestions on where to get started?

**Reply by MarioSwitch** [reply]
Traffic in Germany and Netherlands aren't available yet. It was a suggestion for 4.9 but it wasn't developed...

**Reply by TobiPeterG** [reply]
> Does anyone know if the Traffic implementation for Germany/Netherlands uses the TomTom API, or something else?  And what are the obstacles to implementing it for USA and other countries?  I'm not an Android dev, but I'm a full-stack/cloud dev and would like to contribute, especially to support availability for the US and make the feature (plugin?) available for optional open/crowd-sourced traffic APIs (such as regional or federated servers based on the OpenTraffic platform that @TobiPeterG suggested).  Any suggestions on where to get started?

Hey, feel free to contribute to my project. :)
There are already some discussions about a possible design/architecture and a first PoC already exists.
Here is the link
https://github.com/TobiPeterG/OSMTraffic

Every helping hand is welcome :)

**Reply by yekibud** [reply]
Ha, thanks for the link, @TobiPeterG .  Last few times I tried OrganicMaps OSMAnd won for my use-cases (I think maybe location search/navigation didn't work as well for me).  But I'll give it a try again and follow your progress on traffic.  TBH, I think I'd probably be more comfortable in the OSMAnd Java code base than OrganicMaps C++ - but neither language is my preference/forte, so not a show-stopper.

**Reply by TobiPeterG** [reply]
> Ha, thanks for the link, @TobiPeterG .  Last few times I tried OrganicMaps OSMAnd won for my use-cases (I think maybe location search/navigation didn't work as well for me).  But I'll give it a try again and follow your progress on traffic.  TBH, I think I'd probably be more comfortable in the OSMAnd Java code base than OrganicMaps C++ - but neither language is my preference/forte, so not a show-stopper.

Don't worry, I want to keep the traffic Server independent of the service that uses it. So that Organic Maps and OSMAnd could use it.
There is a discussion about using "Traff' as the data format to exchange the data, there is a gitlab repo with different existing projects, so I would like to adapt it and I have contact with the author of it.
So feel free to work on the OSMand side, the traffic Server side or whatever, as I said, every contribution is absolutely awesome :)

**Reply by yekibud** [reply]
For sure, the server should be independent, but I assume there will need to be some client side UI work for uploading user data and providing route choices.  In any case, definitely a lot of space for collaboration on the server side.  Thanks for the tip about Traff.  Will hopefully be able to find some time over the next month or two to figure out where to dive in.

**Comment by MarioSwitch** [comment]
Just played around with TomTom APIs and get this result: all traffic infos in a specific area (lon/lat of corners) with colors for traffic jams. I've also sorted table by late amount 😊

![image](https://github.com/user-attachments/assets/ab4e6b31-29b3-4b5a-95ad-993f3ebd16d1)

**Comment by sjvudp** [comment]
I think the point is that such API may change at any time, breaking the app. For example in Adobe Lightroom 5 they use some Google map API. At some point in time Google dropped the API, and the map in Lightroom stopped working. As Adobe did not provide an update, the function does not work any more. OK, with periodic updates, you could fix such situations, but in general a mostly self-contained app would be preferable.

**Reply by MarioSwitch** [reply]
If TomTom API doesn't work for any reason, OsmAnd can handle this failure and always work without traffic updates (as now) 🤔

**Comment by mmediaman** [comment]
Can anyone here advise what the legitimate URL is for the TomTom **Maven** artifact repo please?

Looking for: **sdk-search, sdk-routing & sdk-location** for latest legitimate working version.

To be used in _**settings.gradle**_ of a Flutter development:

`dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.PREFER_SETTINGS) 
    repositories {
        google()
        mavenCentral()
        maven {
            url 'https://repositories.tomtom.com/artifactory/maven'
        }`
